nacos/src/test/java/com/linecorp/armeria/internal/nacos/NacosTestBase.java

```diff
@@ -0,0 +1,152 @@
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.CompletionActions;
+import com.linecorp.armeria.internal.testing.FlakyTest;
+import com.linecorp.armeria.server.AbstractHttpService;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A helper class for testing with Nacos.
+ */
+@FlakyTest
+@Testcontainers(disabledWithoutDocker = true)
+public abstract class NacosTestBase {
+    protected static final String serviceName = "testService";
+    protected static final String NACOS_AUTH_TOKEN = "armeriaarmeriaarmeriaarmeriaarmeriaarmeriaarmeriaarmeria";
+    protected static final String NACOS_AUTH_SECRET = "nacos";
+
+    protected static List<Endpoint> newSampleEndpoints() {
+        final int[] ports = unusedPorts(3);
+        return ImmutableList.of(Endpoint.of("host.docker.internal", ports[0]).withWeight(2),
+                                Endpoint.of("host.docker.internal", ports[1]).withWeight(4),
+                                Endpoint.of("host.docker.internal", ports[2]).withWeight(2));
+    }
+
+    @Container
+    static final GenericContainer nacosContainer =
+            new GenericContainer(DockerImageName.parse("nacos/nacos-server:v2.3.0-slim"))
+                    .withExposedPorts(8848)
+                    .withEnv("MODE", "standalone")
+                    .withEnv("NACOS_AUTH_ENABLE", "true")
+                    .withEnv("NACOS_AUTH_TOKEN", NACOS_AUTH_TOKEN)
+                    .withEnv("NACOS_AUTH_IDENTITY_KEY", NACOS_AUTH_SECRET)
+                    .withEnv("NACOS_AUTH_IDENTITY_VALUE", NACOS_AUTH_SECRET);
+
+    @Nullable
+    private static URI nacosUri;
+
+    protected NacosTestBase() {}
+
+    @BeforeAll
+    static void start() {
+        // Initialize Nacos Client
+        nacosUri = URI.create(
+                "http://" + nacosContainer.getHost() + ':' + nacosContainer.getMappedPort(8848));
+    }
+
+    protected static NacosClient client(@Nullable String serviceName, @Nullable String groupName) {
+        final NacosClientBuilder builder;
+        if (serviceName != null) {
+            builder = NacosClient.builder(nacosUri, serviceName);
+        } else {
+            builder = NacosClient.builder(nacosUri, NacosTestBase.serviceName);
+        }
+
+        if (groupName != null) {
+            builder.groupName(groupName);
+        }
+        return builder.authorization(NACOS_AUTH_SECRET, NACOS_AUTH_SECRET)
+                .build();
+    }
+
+    protected static URI nacosUri() {
+        checkState(nacosUri != null, "nacosUri has not initialized.");
+        return nacosUri;
+    }
+
+    protected static int[] unusedPorts(int numPorts) {
```

Comment:
To other reviewers) I guess the ideal approach (binding to port 0) can be done for all similar tests at once. I think this code is fine if the CI passes reliably at least for now.